sdstore/md5helper.go

```diff
@@ -0,0 +1,93 @@
+// Taken and modified from https://blog.golang.org/pipelines
+
+package sdstore
+
+import (
+	"crypto/md5"
+	"encoding/hex"
+	"errors"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+)
+
+// A result is the product of reading and summing a file using MD5.
+type result struct {
+	path string
+	sum  string
+	err  error
+}
+
+// sumFiles starts goroutines to walk the directory tree at root and digest each
+// regular file.  These goroutines send the results of the digests on the result
+// channel and send the result of the walk on the error channel.  If done is
+// closed, sumFiles abandons its work.
+func sumFiles(done <-chan struct{}, root string) (<-chan result, <-chan error) {
+	// For each regular file, start a goroutine that sums the file and sends
+	// the result on c.  Send the result of the walk on errc.
+	c := make(chan result)
+	errc := make(chan error, 1)
+	go func() {
+		var wg sync.WaitGroup
+		err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			if !info.Mode().IsRegular() {
+				return nil
+			}
+			wg.Add(1)
+			go func() {
+				data, err := ioutil.ReadFile(path)
```

Comment:
To read all the files into memory may not work if there is no enough memory. By default, a build only has 2G and if the memory is not released from the previous build steps, it will be less.

Reply:
It looks like we don't need to read the whole file: https://github.com/golang/go/issues/17225
We can do something like this: 
- https://tip.golang.org/pkg/crypto/md5/#example_New_file
- http://mrwaggel.be/post/generate-md5-hash-of-a-file-in-golang/